kiwi_keg/generator.py

```diff
@@ -153,6 +155,81 @@ def create_custom_scripts(self, override: bool = False):
+                            for overlay_path in overlay_include:
+                                overlay_full_path = os.path.join(
+                                    self.image_definition.overlay_root,
+                                    overlay_path
+                                )
+                                # loop all the file paths of overlay sub directory 'overlay_path'
+                                for name in KegUtils.get_all_files(overlay_full_path):
+                                    rel_path = os.path.relpath(name, overlay_full_path)
+                                    new_dir = os.path.dirname(rel_path)
+                                    if new_dir:
+                                        new_dir = os.path.join(overlay_dest_dir, new_dir)
+                                        os.makedirs(new_dir, exist_ok=True)
+                                    dest_file = os.path.join(overlay_dest_dir, rel_path)
+                                    shutil.copy(name, dest_file)
+
+                            if tar_overlays:
+                                overlay_tarball_name = '{}.tar.gz'.format(overlay_name)
+                                self._create_tarball(
+                                    os.path.join(self.dest_dir, overlay_tarball_name),
+                                    overlay_dest_dir
+                                )
+                                shutil.rmtree(overlay_dest_dir)
+                                self._update_config_kiwi(
+                                    overlay_tarball_name,
+                                    overlay_dest_dir
+                                )
+
+            if tar_overlays and not has_overlays:
+                log.warn(
+                    'Attempt to create a tarball but not overlay paths were provided.'
+                )
+
+    @staticmethod
+    def _create_tarball(tarball_dir, dest_dir):
+        with tarfile.open(tarball_dir, 'w:gz') as tar:
+            for overlay_dir in os.scandir(dest_dir):
+                tar.add(
+                    overlay_dir.path,
+                    arcname=overlay_dir.name
+                )
+
+    def _update_config_kiwi(self, archive_name, dest_dir):
+        if os.path.exists(dest_dir) and os.path.exists(self.kiwi_description):
+            etree_parser = etree.XMLParser(remove_blank_text=True)
+            kiwi_xml = etree.parse(self.kiwi_description, parser=etree_parser)
+            kiwi_root = kiwi_xml.getroot()
+
+            image_element = kiwi_root.find('.//packages[@type="image"]')
+
+            archive_element = etree.SubElement(image_element, 'archive')
+            archive_element.attrib['name'] = archive_name
+            tree = etree.ElementTree(kiwi_root)
+            tree.write(
+                self.kiwi_description,
+                encoding="utf-8",
+                xml_declaration=True,
+                pretty_print=True
+            )
+
```

Comment:
If we decide to manage the XML content by keg outside of the template the best way to do it is by using the KIWI API

```python
import sys
import logging

from kiwi.xml_description import XMLDescription
from kiwi.xml_state import XMLState
from kiwi import xml_parse

description = XMLDescription('/home/ms/Project/kiwi-descriptions/suse/x86_64/suse-leap-15.2/config.xml')

xml_data = description.load()

xml_state = XMLState(
    xml_data=xml_data, profiles=['Disk'], build_type='oem'
)

package_section = xml_state.get_packages_sections(
    section_types = ['image']
)[0]

package_section.add_archive(
    xml_parse.archive(name='foo.tar.gz')
)

xml_data.export(
    outfile=sys.stdout, level=0
)
```

be aware this will bring back the comment issue

Reply:
because of the side effects when using the kiwi api (comments and auto update to the schema of the api) I suggest to handle the XML settings via the schema templates as it is done for the other settings as well